### PR TITLE
Track validation errors on remaining provider interface controllers

### PR DESCRIPTION
**Pull request by steventux**
## Context

We've added a way to track validation errors from Manage and Apply. So now we need to make various controller actions track any validation errors.

## Changes proposed in this pull request

Tracks validation errors from Notes form.
Tracks validation errors from various offers controller actions.

## Link to Trello card

https://trello.com/c/ROJfcasA/3609-track-validation-errors-on-manage

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

File: spec/requests/provider_interface/offers_controller_spec.rb [hunk 1]
```diff
@@ -68,4 +68,63 @@
...
+    before { allow_any_instance_of(ProviderInterface::OfferWizard).to receive(:valid_for_current_step?).and_return(false) }
+
+    it 'tracks errors on create' do
+      application_choice.update(status: 'awaiting_provider_decision')
+      expect {
+        post provider_interface_application_choice_offer_path(application_choice)
+      }.to change(ValidationError, :count).by(1)
+    end
+
+    it 'tracks errors on (providers) update' do
```
Comment (benswannack) [inline spec/requests/provider_interface/offers_controller_spec.rb:91]: We don't currently cover the create actions with these specs.

Could we commonize this sort of spec into a shared example so these files look more like a list of endpoints which should track validation errors?

Comment (steventux) [reply spec/requests/provider_interface/offers_controller_spec.rb]: :+1: Thanks @benswannack , I've added missing create actions (my routes grepping let me down there). Not a huge fan of shared examples outside the spec itself as it adds indirection so I've amended to add a shared example in the offers controller spec common to the offers and sub contollers actions (ie not common across all similar controller specs).

File: spec/requests/provider_interface/offers_controller_spec.rb [hunk 2]
```diff
@@ -68,4 +74,145 @@
...
+    let(:trait) { :offer }
+    let(:application_choice) do
+      create(
+        :application_choice, trait,
+        application_form: application_form,
+        course_option: course_option
+      )
+    end
+
+    before { allow_any_instance_of(ProviderInterface::OfferWizard).to receive(:valid_for_current_step?).and_return(false) }
```
Comment (benswannack) [inline spec/requests/provider_interface/offers_controller_spec.rb:89]: As discussed, can we avoid doing the `allow_any_instance_of`?

**Review by benswannack — COMMENTED** [review]
Looking good, we've got good spec coverage now :)

File: spec/support/stub_models_with_errors.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+def stub_model_instance_with_errors(klass, attrs = {}, error_messages = {})
```
Comment (benswannack) [inline spec/support/stub_models_with_errors.rb:1]: This is nice! Could we put it in a module and only include it where necessary?

**Review by benswannack — APPROVED** [review]
LGTM!

**Pull request merged**